components/tools/OmeroPy/src/omero/plugins/metadata.py

```diff
@@ -554,6 +565,87 @@ def populateroi(self, args):
+            'm': UnitsLength.METER,
+            'km': UnitsLength.KILOMETER
+        }
+
+        if args.unit not in units:
+            self.ctx.die(100, "%s not recognized as valid unit." % args.unit)
+
+        if not args.x and not args.y and not args.z:
+            self.ctx.die(100, "No pixel sizes specified.")
+
+        if md.get_type() == "Screen":
+            q = """SELECT pix FROM Pixels pix, WellSample ws, Plate p,
+                   ScreenPlateLink spl WHERE
+                   spl.child=p AND pix.image=ws.image AND ws.well.plate=p AND
+                   spl.parent.id=:id"""
+        elif md.get_type() == "Plate":
+            q = """SELECT pix FROM Pixels pix, WellSample ws WHERE
+                   pix.image=ws.image AND ws.well.plate.id=:id"""
+        elif md.get_type() == "PlateAcquisition":
+            q = """SELECT pix FROM Pixels pix, WellSample ws WHERE
+                   pix.image=ws.image AND ws.plateAcquisition.id=:id"""
+        elif md.get_type() == "Well":
+            q = """SELECT pix FROM Pixels pix, WellSample ws WHERE
+                   pix.image=ws.image AND ws.well.id=:id"""
+        elif md.get_type() == "Project":
+            q = """SELECT pix FROM Pixels pix, DatasetImageLink dil,
+                   ProjectDatasetLink pdl WHERE dil.child=pix.image AND
+                   dil.parent=pdl.child AND pdl.parent.id=:id"""
+        elif md.get_type() == "Dataset":
+            q = """SELECT pix FROM Pixels pix, DatasetImageLink dil WHERE
+                   dil.child=pix.image AND dil.parent.id=:id"""
+        elif md.get_type() == "Image":
+            q = """SELECT pix FROM Pixels pix WHERE pix.image.id=:id"""
+        else:
+            raise Exception("Not implemented for type %s" % md.get_type())
+
+        params = omero.sys.ParametersI()
+        params.addId(md.get_id())
+        pixels = client.getSession().getQueryService()\
+                                    .findAllByQuery(q, params)
+        for pixel in pixels:
+            if args.x:
+                pixSize = pixel.getPhysicalSizeX()
+                if not pixSize:
+                    pixSize = omero.model.LengthI(args.x, units[args.unit])
+                    pixel.setPhysicalSizeX(pixSize)
+                else:
+                    pixSize.setValue(args.x)
+                    pixSize.setUnit(units[args.unit])
+
+            if args.y:
+                pixSize = pixel.getPhysicalSizeY()
+                if not pixSize:
+                    pixSize = omero.model.LengthI(args.y, units[args.unit])
+                    pixel.setPhysicalSizeY(pixSize)
+                else:
+                    pixSize.setValue(args.y)
+                    pixSize.setUnit(units[args.unit])
+
+            if args.z:
+                pixSize = pixel.getPhysicalSizeZ()
+                if not pixSize:
+                    pixSize = omero.model.LengthI(args.z, units[args.unit])
+                    pixel.setPhysicalSizeZ(pixSize)
+                else:
```

Comment:
Are these `else` worthwhile or simpler just to always do the first clause? (Can instantiate the `LengthI` outside the loop so the pixels share it?)

Reply:
Always create a new `LengthI` object? Is the old one automatically removed from the database then?

Reply:
Yes, definitely. They are `structs`.

Reply:
They don't exist in the database as first-class objects.